### PR TITLE
Use font-kit to implement XeTeXFontInst

### LINKED ISSUE #142: Merge xetex_layout_engine.rs into xetex_layout_interface.rs
They used be the same cpp source code file. The `xetex_layout_engine.rs` is the header part in the first batch of transpilation, and `xetex_layout_interface.rs` is the actual c source part. They should be merged together.

### PR DESCRIPTION
**Pull request by cormacrelf**
#186 It's a work in progress. Notable semi-related things included:

* The Read implementation was converting `-1` (error) to u64 and telling
read_to_end to start reading at the other end of the universe.
* Many cleanups of the font manager/font info c2rust conversion.
* fixes #142
* FT_MAKE_TAG macro replaced
* Some name splitting code refactored, to make `find_native_font` more manageable as font-kit will hopefully replace some of it.
* Fixed some non-FFI-safe warnings that are new in recent nightly compilers I think?

Some things that need work:

* The global `name_of_file` and other tex globals are way too involved in font loading. It's painfully unnecessary. set/getReqEngine also.
* There's heavy mixing of the slash-code font loading options like "/AAT", with the rest of the font loading. I'm not quite sure how that works or how it could be disentangled. Also, font-kit will find a font by name, but if it has to search by name and by type... ?
* PlatformFontRef has replaced XeTeXFont, which was just a type alias, but that has made it clear that not every PlatformFontRef is actually a PlatformFontRef -- many are cast willy-nilly to XeTeXFontInst and the platform-specific subclasses. This is terrible.
* The C++ inheritance needs to go. I haven't come up with a neat way of doing it yet.
* The harfbuzz interactions could be done with the `user_data` pointing to a font_kit Font. I'm not sure Font supports all the things harfbuzz is going to need, but you can always keep forcing font-kit to use the freetype loader and use Font.native_font() to access FT_Face.
* I think putting things in harfbuzz user_data should use a Pin pointer?

Some notes:

* loadAATFont and loadOTFont return completely different objects, each under a `*mut libc::c_void` pointer, but both pointers are stored in the `font_layout_engine` static mut. `font_area` 's data stored at the same offset is effectively an enum tag of layout engine types.
* loadAATFont's "layout engine" is actually just a CFDictionaryRef of font attributes. Every thing the "layout engine" is required to do, xetex_ext calls into some aat:: function for, after discovering font_area is `0xffffu32`. So you could think of it as a layout engine, since the resulting API looks similar to using functions (not methods) from xetex_layout_engine.
* loadOTFont is poorly named. It doesn't load the font. It takes a font, matches some options, and essentially wraps `createLayoutEngine`. No other code uses `createLayoutEngine`.
* loadOTFont returns a XeTeXLayoutEngine, which holds a `font: XeTeXFontInst` from `createFont`, and `fontRef: PlatformFontRef`, which is the font loaded with `findFontByName`.
* I honestly have no idea why you don't see panics in `measure_native_node`, which doesn't handle AAT fonts but is called during, among other places, line_break.

Here's the big plan:

* Make a trait for everything in xetex_ext that checks `font_area`/`0xffffu32`/`0xfffeu32`
* `impl LayoutEngine` for a newtype `struct AATLayoutEngine { attributes: CFDictionaryRef }`
* `impl LayoutEngine for XeTeXLayoutEngine`
* Store `Box<dyn LayoutEngine>` in font_layout_engine, and delete font_area. (Hooray!)
* See if you can use Skribo to make another `impl LayoutEngine` similar to `XeTeXLayoutEngine`
* Improve the LayoutEngine API over time

### CODE REVIEW

File: engine/src/xetex_font_info.rs [hunk 1]
```diff
@@ -1877,48 +333,86 @@ unsafe extern "C" fn xbasename(mut name: *const libc::c_char) -> *const libc::c_
...
+        XeTeXFontInst_initialize(self, pathname, index)
+    }
+
+    pub unsafe extern "C" fn units_to_points(&self, units: f32) -> f32 {
+        (units * self.m_pointSize) / (self.m_unitsPerEM as f32)
+    }
+
+    pub unsafe fn get_glyph_bounds(&self, gid: GlyphID) -> GlyphBBox {
+        if let Some(font) = &self.fk_font {
+            if let Ok(rect) = font.typographic_bounds(gid as u32) {
```
Comment (cormacrelf) [inline engine/src/xetex_font_info.rs:349]: FYI, this is the only place we're actually querying font information so far. Further down, font-kit is also used to load the FT_Face.

File: engine/src/xetex_font_manager.rs [hunk 2]
```diff
@@ -1872,137 +529,93 @@ pub unsafe extern "C" fn XeTeXFontMgr_findFont(
...
-                cp,
-                b"AAT\x00" as *const u8 as *const libc::c_char,
-                3i32 as libc::c_ulong,
-            ) == 0i32
-            {
+    if let Some(variant_string) = variant {
+        let mut varString = String::new();
+        let mut slice = &variant_string[..];
+        while !slice.is_empty() {
+            if slice.starts_with("AAT") {
```
Comment (cormacrelf) [inline engine/src/xetex_font_manager.rs:534]: This is the slash-options intermingling I was talking about.

**Comment by cormacrelf** [comment]
Update:

* Swapped out the void*-and-font_area-based dynamic typing for enums, using enum_dispatch to allow dropping down to non-trait methods.
* Had an annoying time initializing a `static mut` hashmap, so it's in a lazy RefCell, which is suboptimal. Rust really doesn't make this bad pattern easy.
* There's a bad measurement going on somewhere.

**Comment by burrbull** [comment]
Maybe if you split this PR on several and rebase it will be much easier to understand what's going on here.

**Comment by burrbull** [comment]
@cormacrelf 
Could you make another try?

**Comment by burrbull** [comment]
Closing this as all main changes are already in `oxidize`

**Pull request closed without merging**